Tizen 7.0 Public M2 Release Note (#1694)
Samsung/tizen-docs: docs/platform/release-notes/tizen-7-0-m2.md

# Tizen 7.0 Public M2

Release date: Oct. 31, 2022

## Release notes

### System (kernel and system framework)

#### New and changed features

- Kernel feature

>   - kfence, a sampling-based memory errobluer detector, support with CONFIG_KFENCE on Raspberry Pi 4

Reviewer: will accept

>   - kfence, a sampling-based memory errobluer detector, support with CONFIG_KFENCE on Raspberry Pi 4.

Reviewer: "safety error" is much better. 
will accept

- Resource Monitor
  - Unified resource monitor framework and capi-system-resource-monitor have been added to monitor the various resource status.

>   - Web-based resource viewer has been added to monitor the whole resource status of Tizen device.

Reviewer: will accept

- Dlog

>   - Zero-copy buffer technique has been developed in order to reduce the printing overhead of log.

Reviewer: will accept

> - Lightweight Multi-user session support

Reviewer: will accept

  - Lightweight multi-user session based on sub-session has been developed.

>   - Directory and file protection of users has been developed.

Reviewer 2: ```suggestion
  - Directory and file protection of users have been developed.

```

Author: Isn't "has" correct expression grammatically as the "protection" is the subject of the sentence?

Reviewer 2: @wtprince,
Yes, you're right, '**file protection**' is the subject but so is '**directory**'. Since we have two subjects in this sentence, we suggest using '**have**' instead of '**has**'.

  - Memory usage and user switching latency have been optimized.
- Device and sensor management

>   - Power state transition callbacks have been developed such as sleep, reboot, poweroff and so on.

Reviewer: will accept

  - Delay function of power transition according to user requirement has been developed.
- Resource management
  - CPU boosting and inheritance C API have been developed and optimized.
  - CPU boosting C# API has been developed for dotnet app boosting.
  - Socket activation for on-demand CPU boosting has been supported.

#### Fixes

- Sensor management
  - Gyroscope rotation vector sensor and gyroscope orientation sensor have been initialized to prevent the wrong operation.

### System (system reliability and OS upgrade)

#### New and changed features

- OS Upgrade and System Recovery

>   - OS upgrade with A/B Partition has been developed to prevent unexpected FOTA failure.

Author: will accept

  - Perl-XML-Simple
    - Version 2.18 has been upgraded to version 2.25.
  - Perl-XML-Twig
    - Version 3.39 has been upgraded to version 3.52.
  - Perl-YAML
    - Version 0.84 has been upgraded to version 1.3.
  - sqlite2
    - Version 3.37.2 has been upgraded to version 3.39.0.
  - wdiff
    - Version 1.1.2 has been upgraded to version 1.2.2.
  - Yaml-cpp
    - Version 0.7.0 has been added.

### Application framework

#### New and changed features

- TIDL
  - TIDL compiler extension for MQTT plugin
    - The TIDL compiler has been updated to generate code that uses the MQTT protocol.
  - Dart programming language support
    - The TIDL compiler has been updated to generate code that uses the Dart programming language.
- Light multi-user system
  - Cross-user sandboxing has been added.

#### Fixes

- Parcel
  - Insert performance has been improved.

### Window and interaction

#### New and changed features

- Text Input Framework
  - The class and APIs related to ElmSharp have been deprecated.
  - The feature of NUI IME has been improved.
    - The UI performance of NUI Reference IME has been improved.
    - Emoticon layout has been supported.

>   - Separates the Setting application from NUI IME to launch it directly by system keyboard setting application.

Reviewer 3: will accept

  - Keyboard setting application has been implemented as NUI.
  - Key code has been provided when pressing or releasing IME key.
- Autofill
  - Autofill setting application has been implemented as NUI.

> - Multi Assistant

Reviewer 2: ```suggestion
- Multi-Assistant

```

Reviewer 4: will accept

  - The new API for sending wakeup word has been added.

>   - The new API for getting whether the voice assistant is enabled.

Reviewer 2: ```suggestion
  - The new API for getting if the voice assistant is enabled.

```

Reviewer 4: will accept

>   - The memory usage of multi assistant has been reduced by not loading unnecessary wakeup engine.

Reviewer 2: ```suggestion
  - The memory usage of multi-assistant has been reduced by not loading unnecessary wakeup engine.

```

Reviewer 4: will accept

- TTS Framework

>   - The C API to provide the activated mode among default, screen reader, notification and etc. to the TTS engine has been added.

Reviewer 4: will accept

>   - The C API to provide the activated mode among default, screen reader, notification, etc. to the TTS engine has been added.

Author: will accept

  - The C/C# API to provide TTS service states such as ready, synthesizing, and playing to TTS clients has been added.
- Voice Control Framework

>   - The IPC interface between the voice-control engine and voice-control clients has been changed from D-Bus to TIDL, which is a proper IPC method between applications.

Reviewer 4: will accept

  - The C API to reduce background volume when the voice manager client wants has been added.
- Multi-Modal Framework
  - The new voice-touch feature has been added.
    - The screen analyzer as one of the input modalities has been added.
    - The voice recognition input modality has been added.

>     - IU (Intent Understanding) Module for the voice touch has been enhanced.

Reviewer 2: ```suggestion
    - IU (Intent Understanding) module for the voice touch has been enhanced.

```

Reviewer 4: will accept

    - Always listening mode has been supported.
- Enlightenment
  - The ratio-fit resize has been added. It provides that a window can be resized with a ratio-fit value.
  - The done event for interactive move and resize has been added. A wayland client can know when the interactive move and resize has been done.

#### Fixes

- Enlightenment

>   - Several issues on Drag and Drop have been fixed.

Reviewer 5: I will accept. Thank you.

    - An issue of hiding a dragged window intermediately has been fixed.
    - An issue of misusing a hash function has been fixed.
    - An issue of sending a drop event when the drop is cancelled has been fixed.

### Graphics and UI

#### New and changed features

- Rendering
  - Support for a dedicated UI thread has been added.
  - Support for EGL_EXT_image_dma_buf_import has been added.
- Scene3D
  - Tizen.NUI.Scene3D namespace has been added.
  - SceneView and Model have been added.
  - Support for facial animation has been added.
  - Support for cubemap images has been added.
  - BVH loader has been added.
- View
  - Various properties about touch and key event have been added.
- Input and Gesture
  - Pan gesture behavior has been optimized.
- Images
  - Support for Lottie dynamic properties has been added.

>   - Support for asynchronous svg file loading has been added.

Reviewer 6: will accept

  - Support for asynchronous Lottie file loading has been added.
  - Support for GPU alpha masking has been added.
- Performance Improvement
  - Some image operations have been optimized.
  - Texture cache logic has been optimized.
  - Building time of a uniform map has been decreased.
  - Signal connection time has been decreased.
  - Vector operation time has been decreased.
  - Text rendering time has been decreased.
- Rive-Tizen

>   - rive-cpp code base to support rive 2.0 API has been updated.

Reviewer 6: will accept

  - Skia based animation renderer has been introduced.
- EFL

>   - Webp animation memory usage has been optimized.

Reviewer 2: ```suggestion
  - WebP animation memory usage has been optimized.

```

Author: will accept

  - Stability has been improved.
- Aurum

>   - Support for intelligent UI Automation has been added.

Reviewer 6: will accept

  - Support for the repeated key event has been added.

>   - Support for the interface to get Text minimum rectangle has been added.

Reviewer 6: will accept

>   - Support for the interface to get text minimum rectangle has been added.

Reviewer 2: I am a bit confused regarding what *text minimum rectangle* means

Reviewer 6: it means the area occupied by rendered texts.

  - Support to get window information from display server has been added.
  - The performance key event response time has been improved.

>   - Support for the interface to get object information with specific area has been added..

Reviewer 6: will accept

>   - Support for the interface to get object information with specific area has been added..

Reviewer 2: Hello @hyunjushin,
It seems this line has been deleted from the `docs/platform/release-notes/tizen-7-0-m2.md` file. 
Is there any specific reason why?

  - Support to get Process ID from AT-SPI has been added.
- UI Analyzer
  - Support for query function of the focusable object has been added.
- Accessibility
  - ATSPI support for embedded UIs in NUI/DALi (widget applications, web content).

>   - New reading UX in screen reader

Author: will accept

>   - New reading UX in screen-reader

Reviewer 6: will accept

>   - New property in NUI/DALI to support automation: DevelControl::Property::AUTOMATION_ID

Reviewer 2: ```suggestion
  - New property in NUI/DALi to support automation: DevelControl::Property::AUTOMATION_ID

```

Reviewer 6: will accept

  - New NUI API for accessibility attributes: View.AccessibilityAttributes
  - Registration of full EFL setting application package name in the AT-SPI registry.
- Figma

>   - Support for converting Figma to xaml has been added.

Reviewer 6: will accept

>   - Figma plugin : `https://github.com/nui-dali/FigmaToNUIXaml`

Reviewer 6: will accept

- Components
  - Support for drag and drop between applications has been added.

#### Fixes

- Various partial update defects have been fixed.
- Visual defect of scrolling text has been fixed.
- Various text defects have been fixed.
- Various ATSPI defects have been fixed.
- Various Aurum and UI analyzer defects have been fixed.

### Multimedia framework

#### New and changed features

- MediaVision

>   - On-Device learning based Face Recognition API.

Reviewer 7: "On-device" is used commonly not "On device". Plese use "On-device" instead.

>   - On-Device learning based Face Recognition API.

Reviewer 8: I will take it. Thanks.

    - Support for face registration in runtime has been added.
    - Support for face unregistration in runtime has been added.
    - Support for face recognition in runtime has been added.
  - 3D API set has been added.

>     - Support for Depth process based on Stereo has been added.

Reviewer 7: Will accept.

    - Support for PointCloud process has been added.
  - Region of Interest tracker set has been added.
- Native WebRTC
  - Encoder bitrate APIs for video source have been added.
  - Transceiver codec APIs have been added.
  - NULL source type for recvonly direction has been added.
  - File path and looping APIs for media file source have been added.

>   - Device id APIs for camera source have been added.

Reviewer 2: ```suggestion
  - Device ID APIs for camera source have been added.

```

Reviewer 8: Will accept

- Sound Manager
  - Deprecated APIs since Tizen 4.0 have been removed.
- Media Streamer
  - All APIs have been deprecated.
- Media Content
  - Pinyin filter keywords have been deprecated.
  - CreateThumbnail C# API has been added and CreateThumbnailAsync C# API has been deprecated.
- Media Camera
  - APIs for network camera and extra stream preview have been added.
- Media Common
  - Memory leak detector feature has been added.
  - The new MediaPacket constructor based on ref count has been added.
  - The Display APIs for ElmSharp have been deprecated.
- Media Util
  - ThumbnailExtractor ExtractAsync has been deprecated.

#### Fixes

- Native WebRTC
  - Validation of STUN/TURN server URL has been improved.
  - Data type of 'packets-lost' issue has been fixed.
  - OPUS channel information has been added to rtpmap attribute of SDP description to comply with Web API.
  - Default resolution for screen source has been fixed to use the display resolution of a target device.
  - An invalid mute operation of camera source not using Tizen memory has been fixed.
- Media Common
  - Daemon launching failure issue related with display service has been resolved.

>   - Deadlock issue by non thread safe Cynara API has been resolved.

Reviewer 2: ```suggestion
  - Deadlock issue by non thread-safe Cynara API has been resolved.

```

Reviewer 8: Will accept

>   - Deadlock issue by non thread safe cynara API has been resolved.

Reviewer 8: I will take it. Thanks.

  - Player data thread memory issue (heap-use-after-free) has been resolved.
- Media Camera
  - The C# crash issue in RPI 64bit system has been fixed.

>     - Changed bool to byte in EncodedPlane internally for 64bit union struct marshaling

Reviewer 9: will accept

### Network and connectivity

#### New and changed features

- Wi-Fi
  - New features for Wi-Fi roaming have been added.

>     - IEEE 802.11k feature has been added for scanning nearby Aps quickly.

Author: will accept

>     - IEEE 802.11 k feature has been added for scanning nearby Aps quickly.

Reviewer 2: ```suggestion
    - IEEE 802.11k feature has been added for scanning nearby APs quickly.

```

Reviewer 10: Will accept

>     - IEEE 802.11 r feature has been added for authenticating quickly while roaming.

Reviewer 10: Will accept

>     - IEEE 802.11 v feature has been added for triggering roaming from the infrastructure AP based on the network status.

Reviewer 2: ```suggestion
    - IEEE 802.11v feature has been added for triggering roaming from the infrastructure AP based on the network status.

```

Reviewer 10: Will accept

  - The data sending issue on LE CoC has been fixed.
  - The wrong operation about same GATT services and descriptors has been fixed.
- Network
  - The time issue about network information change has been fixed.
  - The issue about BSS (Basic Service Set) transition has been fixed.

### Service framework

#### Fixes

- Account Framework
  - Potential defects have been fixed.
- Sync-Manager
  - Potential defects have been fixed.
- Location Framework
  - Potential defects have been fixed.
- Messaging Framework
  - Potential defects have been fixed.

### Web framework

#### New and changed features

- Web Engine
  - Tizen.WebView C# API has been deprecated.
  - Resource size (content_shell.pak) has been optimized.
  - Launching performance and memory have been improved.
- Web Runtime
  - Open source Node.js 14.15.4 version has been applied.

#### Fixes

- Web Engine
  - Accessibility defects have been fixed.
  - NUI WebView related defects have been fixed.
  - ewk_view_scroll_by defect has been fixed.

### Lightweight web solution

#### New and changed features

- Lightweight Web Engine
  - Web Engine
    - CSS SVGRadialGradientElement has been added.

>   - Javascript Engine

Reviewer 2: ```suggestion
  - JavaScript Engine

```

Reviewer 11: will accept

    - Private instance fields/methods of class have been added.
    - Top-level await feature has been added.

>     - RegExp match indeces has been added.

Reviewer 2: ```suggestion
    - RegExp match indices have been added.

```

Reviewer 11: will accept

- JS based lightweight backend service FW
  - Lightweight Node.js
    - Tizen device API module has been added.
    - Binding Glib message queue has been added.
    - Memory tracing APIs have been added.
    - App templates for RPMs and TPKs have been added.

>     - Preliminary Javascript debugging via VS code has been added.

Reviewer 2: ```suggestion
    - Preliminary JavaScript debugging via VS code has been added.

```

Reviewer 11: will accept

#### Fixes

- Lightweight Web Engine
  - SVG defects have been fixed.
  - Flex layout defects have been fixed.

>   - Animated gif defects have been fixed.

Reviewer 2: ```suggestion
  - Animated GIF defects have been fixed.

```

Reviewer 11: will accept

- Lightweight Node.js

>   - Visualizing stack traces has been fixed.

Author: will accept

  - Thread-safety defects have been fixed.

>   - App Template defects have been fixed.

Reviewer 2: ```suggestion
  - App template defects have been fixed.

```

Reviewer 11: will accept

### Tizen .NET

#### New and changed features

- .NET
  - .NET Multi-platform App UI (MAUI)
    - .NET MAUI General Availability (GA) has been supported.
  - TizenFx
    - Non-intuitive key focus events of Tizen.NUI have been deprecated.
    - IsLooping method has been added to MediaFileSource class.
    - New construct has been added to MediaPacket class.

>     - Encoder/Decoder for Jpeg XL image have been added to Tizen.Multimedia.Util.

Reviewer 2: ```suggestion
    - Encoder/Decoder for JPEG XL image has been added to Tizen.Multimedia.Util.

```

Reviewer 12: Will accept

    - TransceiverCodec has been added to Tizen.Multimedia.Remoting.
    - Unused Camera class and CameraType of NUI have been deprecated.
    - Deep learning face recognition has been added to Tizen.Multimedia.Vision.
    - ROI tracker has been added to Tizen.Multimedia.vision.
    - CameraDeviceId and EncoderBitrate have been added to Tizen.Multimedia.Remoting.
    - CreateThumbnail has been added to Tizen.Content.MediaContent.
    - ElmSharp and related constructors have been deprecated.
    - ExtractAsync of ThumbnailExtractor class has been deprecated.
    - Deprecated method of AudioManager has been removed.
    - Tizen.WebView has been deprecated.
    - EncodedPlane has been added to Tizen.Multimedia Camera.
    - Camera and SceneView class has been added to Tizen.NUI.Scene3D.
    - Privilege level of MediaScreenSource class has been changed from public to platform.

### Toolchain

#### New and changed features

- GCC
  - Timezone data has been updated to tzcode 2020a.

#### Fixes

- GCC
  - C++ bugfix patches have been applied (PR c++/61414).
  - Memcmp bugfix patches have been applied (PR middle-end/95189, 95886).
- Glibc

>   - TLS(Thread Local Storage) bugfix patches have been applied (BZ #25051).

Reviewer 2: ```suggestion
  - TLS (Thread Local Storage) bugfix patches have been applied (BZ #25051).

```

Reviewer 13: will accept.

### Machine learning

#### New and changed features

- Machine Learning (ML) API updates
  - ML.Service API Updates

>     - Home Edge-AI and Among-Device AI support: applications may register and deploy AI service modules, which can be accessed by other applications in the device or connected devices. Such Edge-AI or Among-Device AI capability is provided by the API set to launch/start/stop/destroy the AI pipeline descriptions, which have been newly added. Developers may provide the specific inference operations as ML service and other applications may use this service via the network. Note that this capability is compatible with SmartThings and Matter via AITT module and devices without NNStreamer (e.g., TizenRT) can be potentially connected by adopting nnstreamer-edge library.

Reviewer 2: ```suggestion
    - Home Edge-AI and Among-Device AI support: applications may register and deploy AI service modules, which can be accessed by other applications in the device or connected devices. Such Edge-AI or Among-Device AI capability is provided by the API set to launch/start/stop/destroy the AI pipeline descriptions, which have been newly added. Developers may provide the specific inference operations as ML service, and other applications may use this service via the network. Note that this capability is compatible with SmartThings and Matter via AITT module, and devices without NNStreamer (e.g., TizenRT) can be potentially connected by adopting nnstreamer-edge library.

```

Author: will accept

>     - Home Edge-AI and Among-Device AI support: applications may register and deploy AI service modules, which can be accessed by other applications in the device or connected devices. Such edge-AI or among-device AI capability is provided by the API set to launch / start / stop / destroy the AI pipeline descriptions, which has been newly added. Developers may provide the specific inference operations as ML service and other applications may use this service via the network. Note that this capability is compatible with SmartThings and Matter via AITT module and devices without nnstreamer (e.g., TizenRT) can be potentially connected by adopting nnstreamer-edge library.

Reviewer 2: ```suggestion
    - Home Edge-AI and Among-Device AI support: applications may register and deploy AI service modules, which can be accessed by other applications in the device or connected devices. Such Edge-AI or Among-Device AI capability is provided by the API set to launch/start/stop/destroy the AI pipeline descriptions, which have been newly added. Developers may provide the specific inference operations as ML service and other applications may use this service via the network. Note that this capability is compatible with SmartThings and Matter via AITT module and devices without NNStreamer (e.g., TizenRT) can be potentially connected by adopting nnstreamer-edge library.

```

Reviewer 14: will accept.

>     - API set to create/request the query service to the ML service instance has been newly added. Applications may create the query for inference offloading and request it to the ML service instance with its data.

Reviewer 2: ```suggestion
    - API set to create/request the query service to the ML service instance has been newly added. Applications may create the query for inference offloading, and request it to the ML service instance with its data.

```

Author: will accept

>     - API set to create / request the query service to the ML service instance has been newly added. Applications may create the query for inference offloading and request it to the ML service instance with its data.

Reviewer 2: ```suggestion
    - API set to create/request the query service to the ML service instance has been newly added. Applications may create the query for inference offloading and request it to the ML service instance with its data.

```

Reviewer 14: will accept

  - ML.Inference API Updates

>     - API set to set / destroy the ML options has been newly added. Developers may use this API to set the various option of the deep neural network framework.

Reviewer 2: ```suggestion
    - API set to set/destroy the ML options has been newly added. Developers may use this API to set the various option of the deep neural network framework.

```

Reviewer 14: will accept

    - Version 2.8.1 of TensorFlow Lite has been supported as default.
  - ML.Training API Updates

>     - Added New Enumerations for layers

Reviewer 2: ```suggestion
    - Added new enumerations for layers

```

Reviewer 15: will accept

      - ML_TRAIN_LAYER_TYPE_ATTENTION
      - ML_TRAIN_LAYER_TYPE_MOL_ATTENTION
      - ML_TRAIN_LAYER_TYPE_LAYER_NORMALIZATION
      - ML_TRAIN_LAYER_TYPE_POSITIONAL_ENCODING

>     - .NET API has been added. Now, It is possible to construct, control, and train a machine learning model in Tizen device with Tizen.ManchineLearning.Trainng .NET Class.

Reviewer 2: ```suggestion
    - .NET API has been added. Now, it is possible to construct, control, and train a machine learning model in a Tizen device with the Tizen.ManchineLearning.Trainng .NET class.

```

Author: will accept

>       - It provides :

Reviewer 2: ```suggestion
      - It provides:

```

Reviewer 15: will accept

        - Tizen.MachineLeanring.Train.Model
        - Tizen.MachineLearning.Train.Layer
        - Tizen.MachineLearning.Train.Optimizer
        - Tizen.MachineLearning.Train.Dataset
        - And more.
    - Training API in WebAPIs has been added for creation and training new models for machine learning on Tizen.
      - It contains:

>         - Layer creattionLayer(LayterType type);
>         - Dataset createFileDataset()
>         - Layer createOptimizer()
>         - Model createModelWithConfiguration()
>         - Model createMode()

Reviewer 2: Line 492 ends with a semi-colon `;`. Should the following lines also end with semi-colon `;`
```suggestion
        - Layer creattionLayer(LayterType type);
        - Dataset createFileDataset()
        - Layer createOptimizer()
        - Model createModelWithConfiguration()
        - Model createMode()

```

Reviewer 15: This needs to be ```creationLayer(LayterType type)```

        - And more.
- NNStreamer updates
  - NNStreamer has been upgraded to version 2.2.
  - New tensor stream type, Float16, has been added.

>   - Edge AI functionality has been added.

Reviewer 2: ```suggestion
  - Edge-AI functionality has been added.

```

Reviewer 14: will accept

    - nnstreamer-edge package has been newly added. It provides interfaces to support data connection and AI offloading feature between edge devices.

>     - NNStreamer and related plugins now support distributed AI inference by using the interfaces for edge AI.

Reviewer 2: ```suggestion
    - NNStreamer and related plugins now support distributed AI inference by using the interfaces for Edge-AI.

```

Reviewer 14: will accept

  - User customized tensorflow2-lite of tensor filter has been supported. Developers may designate user-supplied tf2-lite instead of the default one.
  - Apache MXNet filter has been newly added.
  - NXP DeepViewRT filter has been newly added.
  - PyTorch has been upgraded to version 1.10.2.

>   - Error messages, exception handling, and documentation have been improved for application / pipeline writers.

Reviewer 2: ```suggestion
  - Error messages, exception handling, and documentation have been improved for application/pipeline writers.

```

Reviewer 14: will accept

  - The Rank limit of tensor stream has been increased from 4 to 8. (Experimental)

>   - edgesrc / edgesink element newly added to support publish-subscribe communication. (Experimental)

Reviewer 2: ```suggestion
  - edgesrc/edgesink element newly added to support publish-subscribe communication. (Experimental)

```

Author: will accept

- NNTrainer updates
  - New layers to support advanced training methods have been added.

>     - Attention, MoL Attention, Multi-Head Attention Layer
>     - Positional Encoding Layer
>     - Layer Normalization Layer
>   - Support for Transformer with Multi-Head Attention Layer has been added.
>   - Memory Optimization Scheme has been improved.
>     - Refactoring Memory Scheduler, Memory Pool, Tensor Pool for better Performance
>     - Consume only 1/3.9 Peak Memory of TFLite v2.9 for VGG Model with better latency (1/1.49)

Reviewer 2: ```suggestion
    - Attention, MoL Attention, Multi-Head Attention Layer.
    - Positional Encoding Layer.
    - Layer Normalization Layer.
  - Support for Transformer with Multi-Head Attention Layer has been added.
  - Memory Optimization Scheme has been improved.
    - Refactoring Memory Scheduler, Memory Pool, Tensor Pool for better Performance.
    - Consume only 1/3.9 Peak Memory of TFLite v2.9 for VGG Model with better latency (1/1.49).

```

Author: will accept

>     - Consume only 1/3.9 Peak Memory of TFLite v2.9 for VGG Model with better Latency (1/1.49)

Reviewer 2: ```suggestion
    - Consume only 1/3.9 Peak Memory of TFLite v2.9 for VGG Model with better latency (1/1.49)

```

>   - Experimental Features (Turned off by Default)

Reviewer 2: ```suggestion
  - Experimental Features (turned off by default)

```

Reviewer 15: will accept

    - Support for Batch-wise Parallelization has been added.
    - Support for Learning Rate Scheduling has been added.
    - Support for Leaky Relu Layer has been added.

>     - Tensorflow-lite Exporter

Reviewer 2: ```suggestion
    - TensorFlow Lite Exporter.

```

Reviewer 15: will accept

    - Support for Gradient Clipping by Global Norm has been added.
    - Scheduled memory swapping to further minimize peak memory consumption, which reduces the peak down to 1/4 of the conventional nntrainer.
- Neural Network Runtime

>   - Runtime has been supported to run nnpackage with multiple models.

Reviewer 2: ```suggestion
  - Runtime has been supported to run NNPackage with multiple models.

```

Reviewer 16: `nnpackage` is a proper notation. Please refer to https://github.com/Samsung/ONE/blob/master/nnpackage/spec/10_packaging_and_manifest.md

  - Conv2D and Depthwise Conv2D support per-channel quantization of uint8 type.
  - TRIX backend supports batch execution which run in parallel with multicore.

#### Fixes

- NNStreamer
  - Several workarounds have been added for glitches in Qualcomm-SNPE's libraries.

> #### Known Issues

Reviewer 2: ```suggestion
#### Known issues

```

Author: will accept

Author: will accept

- NNStreamer
  - Python2 filter has been removed. Only Python3 is supported.
  - Converter and decoder with Python3 codes do not support multi-thread methods. However, multithreading in Python3 filters are now supported.
- NNTrainer

>   - Further memory saving with swapping is experimental and it does not support multi-threaded training.

Reviewer 2: ```suggestion
  - Further memory saving with swapping is experimental, and it does not support multi-threaded training.

```

Author: will accept
